### ISSUE TITLE
Memory leak after upgrading to v2.1.0

### ISSUE THREAD

**Issue by maxpain** [hasura/graphql-engine#7936]
### Version Information

I've upgraded graphql-engine from `v2.0.10` to `v2.1.0`.

### Environment
Google Cloud Platform: Kubernetes Engine

### What is the current behaviour?
Memory usage grows over time (possible memory leak)

### Screenshots or Screencast

<img width="1077" alt="image" src="https://user-images.githubusercontent.com/5875582/146183894-3f510a4a-3ea8-4d76-9166-b789102b1914.png">

**Comment by maxpain** [comment]
<img width="1272" alt="Screenshot 2021-12-15 at 16 07 04" src="https://user-images.githubusercontent.com/5875582/146191929-8fa8d87d-7fd5-4ffe-8fad-4523c8092ab5.png">

CPU Usage on Postgres is also high after upgrade

**Comment by maxpain** [comment]
<img width="1229" alt="image" src="https://user-images.githubusercontent.com/5875582/146193480-2763d4b8-6bb6-48dd-9b0f-0d9eb8fe1f52.png">

**Comment by jberryman** [comment]
@maxpain does the spike at 13:00 represent the time you upgraded to v2.1.0? It looks like requests may be arriving at hasura faster than they can be serviced within postgres (since they are quite slow). So I'd like to understand if this is e.g.

- heavy load on your end
- load that triggers a slow query we can fix in hasura but is unrelated to a regression in v2.1.0
- a pg query that got slower in v2.1.0

**Comment by maxpain** [comment]
> @maxpain does the spike at 13:00 represent the time you upgraded to v2.1.0? 
Yes.

**Comment by jberryman** [comment]
Gotcha, can you share a few of those newly-slow queries with us privately please? (like the green and red bands from the "Total query time" graph) brandon@hasura.io

If you're able to do an EXPLAIN ANALYZE against your db for those, that would be helpful as well.
If you can share a db dump with row data removed and your hasura metadata that also is helpful.

**Comment by maxpain** [comment]
> Gotcha, can you share a few of those newly-slow queries with us privately please? (like the green and red bands from the "Total query time" graph) [brandon@hasura.io](mailto:brandon@hasura.io)
> 
> If you're able to do an EXPLAIN ANALYZE against your db for those, that would be helpful as well. If you can share a db dump with row data removed and your hasura metadata that also is helpful.

Done

**Comment by jberryman** [comment]
Thanks, received and I can see some slow queries. Are you able to do an `EXPLAIN ANALYZE`?

And did 13:00 represent a deployment of new application code or anything else? Or was the hasura version the only thing that changed?

**Comment by maxpain** [comment]
> Are you able to do an EXPLAIN ANALYZE?

Yes, I will compare generated SQL queries and do `EXPLAIN ANALYZE` on both `v2.1.0` and `v2.0.10` versions on the same dataset.

> And did 13:00 represent a deployment of new application code or anything else?

No

> Or was the hasura version the only thing that changed?

Yes

**Comment by maxpain** [comment]
A noticed that only one very big query is affected by that upgrade.

**Comment by maxpain** [comment]
@jberryman I've sent you `explain analyze` results of the same query running on `v2.1.0` against `v2.0.10`

**Comment by maxpain** [comment]
`v2.1.0` has bug with LIMIT when using Object relationships

<img width="1194" alt="Screenshot 2021-12-16 at 00 26 07" src="https://user-images.githubusercontent.com/5875582/146267458-9f198099-8135-4f79-8a7c-669b2ac01034.png">

<img width="1155" alt="Screenshot 2021-12-16 at 00 26 14" src="https://user-images.githubusercontent.com/5875582/146267462-dd807dc8-c812-4fca-b35e-8ed568a0292a.png">

**Comment by tirumaraiselvan** [comment]
@maxpain Just to confirm, this is using the `hasura/graphql-engine:v2.1.0` image and is the open source self-hosted version (not cloud or pro) right?

**Comment by maxpain** [comment]
> @maxpain Just to confirm, this is using the `hasura/graphql-engine:v2.1.0` image and is the open source self-hosted version (not cloud or pro) right?

This bug exists on both open source and cloud versions of v2.1.0.

**Comment by tirumaraiselvan** [comment]
@maxpain That's fine. We are trying to repro with your metadata. Are you running on open source or cloud or both?

**Comment by maxpain** [comment]
> @maxpain That's fine. We are trying to repro with your metadata. Are you running on open source or cloud or both?

I already did reproduction on Hasura Cloud with minimal schema. Please check 'query-limit-bug-test' project if you have an access. Or I can invite you.

Then please execute this query:

```
{
  users(
    where: {id: {_eq: 1}}
    limit: 1
  ) {
    id
    isMeBlacklisted {
      targetUserID: target_user_id
    }
  }
}
```

**Comment by tirumaraiselvan** [comment]
@maxpain Pls do invite me `tiru@hasura.io`

**Comment by maxpain** [comment]
> @maxpain Pls do invite me `tiru@hasura.io`

Done

**Comment by tirumaraiselvan** [comment]
Thanks, pasting repro artifacts here:

[metadata.json.txt](https://github.com/hasura/graphql-engine/files/7725599/metadata.json.txt)
[dump.sql.txt](https://github.com/hasura/graphql-engine/files/7725600/dump.sql.txt)

Query (with user role or admin role): 

```
{
  users(
    where: {id: {_eq: 1}}
    limit: 1
  ) {
    id
    isMeBlacklisted {
      targetUserID: target_user_id
    }
  }
}
```

**Comment by maxpain** [comment]
@tirumaraiselvan, when it will be fixed?

**Comment by maxpain** [comment]
@tirumaraiselvan Could you please provide some information about the fix?

**Comment by tirumaraiselvan** [comment]
@maxpain Really sorry for the delay here. We did triage this internally but missed to update here.

There are 2 aspects to the current "bug":

1. The issue only happens for a manual object relationship. Object relationships are typically meant for a n:1 or a 1:1 relationship i.e. the right hand side relation returns at max one row for a given left hand side row. But, in your case there are multiple rows on the right hand side for the same row. In older versions of Hasura, this would return any one row (randomly selected) for the relationship. It's not clear if this is intentional since an array relationship would be better suited here.

2. We will fix the behaviour to return a single randomly selected row in such a scenario (older behaviour) soon. But the priority of this is not urgent since this is a rather unexpected configuration.

**Comment by sassela** [comment]
> We will fix the behaviour to return a single randomly selected row in such a scenario (older behaviour) soon. But the priority of this is not urgent since this is a rather unexpected configuration.

@maxpain thanks for your patience with this. This change is now available in the [pre-release version v2.3.0-beta.1](https://github.com/hasura/graphql-engine/releases/tag/v2.3.0-beta.1)
